firedrake/mesh.py

```diff
@@ -816,11 +891,52 @@ def make_offset(self, entity_dofs, ndofs):
                     dof_offset[i] = entity_offset[b]
         return dof_offset
 
-    @property
+    @utils.cached_property
     def layers(self):
         """Return the number of layers of the extruded mesh
         represented by the number of occurences of the base mesh."""
-        return self._layers
+        if self.variable_layers:
+            raise ValueError("Can't ask for mesh layers with variable layers")
+        else:
+            return self.cell_set.layers
+
+    @utils.cached_property
+    def variable_layers(self):
+        """Does the extruded mesh have variable layer numbers?"""
+        return not self.cell_set.constant_layers
+
+    def entity_layers(self, height, label=None):
+        """Return the number of layers on each entity of a given plex
+        height.
+
+        :arg height: The height of the entity to compute the number of
+           layers (0 -> cells, 1 -> facets, etc...)
+        :arg label: An optional label name used to select points of
+           the given height (if None, then all points are used).
+        :returns: a numpy array of the number of layers on the asked
+           for entities (or a single layer number for the constant
+           layer case).
+        """
+        if self.variable_layers:
+            return extnum.entity_layers(self, height, label)
+        else:
+            return self.cell_set.layers
+
+    @utils.cached_property
+    def layer_extents(self):
+        """The layer extents for all mesh points.
+
+        For variable layers, the layer extent does not match those for cells.
+
+        :returns: numpy array of layer extents (in PyOP2 format
+            :math:`[start, stop)`), of shape ``(num_cells, 4)`` where
+            the first two extents are used for allocation and the last
+            two for iteration.
+        """
+        extents = extnum.layer_extents(self)
+        if np.any(extents[:, 3] - extents[:, 2] <= 0):
+            raise NotImplementedError("Vertically disconnected cells unsupported")
```

Comment:
Consider this mesh:
```
4

3       +----+
        |    |
2       +----+

1  +----+
   |    |
0  +----+
```

The cell layer extents will be:

```
[0, 2, 0, 2]
[2, 4, 2, 4]
```

The interior facet layer extent will be:

```
[0, 4, 2, 2]
```

Because the latter two entries are respectively `max(0, 2)` and min(2, 4)`.  A difference `<=0` implies that the column of entities has a gap in it because the starting iteration layer is above the stop iteration layer.